Redmine #7030: Don't hang when operating on fifos

worr opened the pull request:
Implemented using `seteuid` and `setegid` per @kacfengine's comments. Had to open new pull request due to some git fudgery on my end. See #2192 for discussion. Related to #2194.

kacfengine commented on libcompat/fchmodat.c:
```diff
@@ -0,0 +1,106 @@
+	uid_t olduid = 0;
+
+    fchmodat_data *data = generic_data;
+    if ((ret = fstatat(data->dirfd, data->pathname, &buf, data->flags)))
```
You should (and in fact, must) use ``fstat`` here, since if the underlying platform doesn't implement ``fstatat``, then it will also use ``generic_at_function`` and get stuck when trying to relock the already held ``CHDIR_LOCK``. Using ``fstat`` is safe though, since you are already inside the correct directory at this point.

kacfengine replied:
Err, not ``fstat``, I mean ``stat`` or ``lstat``, depending on the value of ``flags``.

kacfengine replied:
Damn, I'm wrong again. This is tricky stuff to get right!

What we need to know is who owns the leaf, regardless of whether that is a link or something else. So always using ``lstat`` here is the correct approach. That means that noone can sneak in a link to a sensitive file, since the link would have to be owned by them, and hence would trigger the ``seteuid()`` below.

In addition, I think you should return error if ``flags`` contains ``AT_SYMLINK_NOFOLLOW``, since you cannot possibly honor it (we don't call this anywhere from our code, but it makes it future proof).

kacfengine commented on libcompat/fchmodat.c:
```diff
@@ -0,0 +1,106 @@
+
+	fprintf(stderr, "%d\n", gid);
+	fputs("setegid\n", stderr);
+	if ((ret = setegid(gid))) {
```
Using ``setegid()`` is not necessary, only the owner can change permissions, regardless of the existing permissions for the group.

kacfengine commented on libcompat/fchmodat.c:
```diff
@@ -0,0 +1,106 @@
+		return -1;
+
+	if (setegid(oldgid))
+		return -1;
```
Same as above, can be removed.

kacfengine commented on libutils/file_lib.c:
```diff
@@ -736,14 +725,27 @@ int safe_chmod(const char *path, mode_t mode)
+    if ((dirfd = safe_open(parent_dir, O_RDONLY)) == -1)
+        goto bad;
+
+    if ((ret = fchmodat(dirfd, leaf, mode, AT_SYMLINK_NOFOLLOW)))
```
``flags`` should be ``0``, not ``AT_SYMLINK_NOFOLLOW``, since this is ``chmod``, not ``lchmod``.

kacfengine replied:
See also comments inside ``fchmodat_inner``.

kacfengine commented on libutils/file_lib.c:
```diff
@@ -453,6 +454,13 @@ int safe_open(const char *pathname, int flags, ...)
+            {
+                close(currentfd);
+                return -1;
+            }
```
I think this should be handled elsewhere, since it is not categorically wrong to open a fifo, it's just in the particular way we have been using it from our ``chown`` implementation. In fact, is this even needed anymore? The new implementations of ``safe_chmod``, ``safe_chown`` and ``safe_lchown`` won't open files anymore.

Opening fifos in general could be desirable, for example using ``readfile`` from policy, when you know that another process is at the other end.

kacfengine commented:
Some issues remain, but I agree with the overall approach, this is the correct way to solve it.

The issues that must be fixed:

* The indentation is all screwed up in ``libcompat/fchmodat.c``. Should be four spaces for each level. Also the coding style requires braces around every if statement, even oneliners.
* There is a lot of debug print statements left over in the same file.
* ``fstatat`` issue in ``fchmodat_inner`` (see separate comments).
* Removing ``setegid`` (see separate comments).
* Switch ``fchmodat`` parameter ``AT_SYMLINK_NOFOLLOW`` to zero (see separate comments).
* Remove FIFO detection (see separate comments).

Thanks for working on this very complicated issue!

jimis commented:
This is complex issue, much more complex than it appears. +1 for diving into this hole :-)

worr commented:
Hey,

So since the CLA is just about worked out, I've updated this diff.

I've rebased it against the current master, as well as addressed all of the other issues. @kacfengine: Can you take another look at this? Thanks!

kacfengine commented on libutils/file_lib.c:
```diff
@@ -650,6 +651,32 @@ int safe_chdir(const char *path)
+        goto bad;
+
+    if ((ret = fchownat(dirfd, leaf, owner, group, flags)))
+        goto bad;
```
Code style (here and several other): We want braces around all if blocks, even one liners. I have fixed this for now, but remember it next time!

kacfengine commented:
This is looking very good. I've fixed some minor issues and resubmitted here: https://github.com/cfengine/core/pull/2390. If that gets merged I will close this one.

kacfengine commented:
Fixed in #2390.
